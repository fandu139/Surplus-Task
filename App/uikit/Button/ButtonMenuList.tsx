import { ICON_ARROW_RIGHT } from '../../assets/icon';
import Colors from '../../theme/colors';
import Icon from '../Icon';
import Text from '../Text';
import React from 'react';
import { View, StyleSheet, TouchableOpacity, StyleProp, ViewStyle, TextStyle } from 'react-native';
import FastImage from 'react-native-fast-image';
import AppStyles from '../../theme/appStyles';
import Divider from '../Divider';

/**
 *
 * @param testID
 * @param accessibilityLabel
 * @param iconLeft
 * @param imageLeft
 * @param iconLeftColor
 * @param iconRight
 * @param iconRightColor
 * @param rightTitle
 * @param title
 * @param onPress
 * @param titleStyle
 * @param style
 * @param useDivider
 * @param useBold
 * @param rightTitleStyle
 * @param headerText
 * @param headerRightText
 * @param description
 * @param disabled
 * @param subtitle
 * @param rightContent
 * @returns {*}
 * @constructor
 */

interface Props {
  testID: string;
  testIDTitle?: string;
  testIDRightTitle?: string;
  accessibilityLabel?: string;
  iconLeft?: string;
  imageLeft?: number;
  iconLeftColor?: string;
  iconRight?: string | null;
  iconRightColor?: string;
  iconRightSize?: number;
  onPressIconRight?: () => void;
  testIDIconRight?: string;
  rightTitle?: string;
  title?: string | React.ReactNode;
  onPress?: () => void;
  titleStyle?: StyleProp<TextStyle>;
  style?: StyleProp<ViewStyle>;
  useDivider?: boolean;
  useBold?: boolean;
  rightTitleStyle?: StyleProp<TextStyle>;
  headerText?: string;
  headerRightText?: string;
  description?: string;
  disabled?: boolean;
  subtitle?: React.ReactNode;
  rightContent?: React.ReactNode;
  rightTitleSelectable?: boolean;
}

const ButtonMenuList: React.FC<Props> = ({
  testID,
  testIDTitle,
  testIDRightTitle,
  accessibilityLabel,
  iconLeft,
  imageLeft,
  iconLeftColor,
  iconRight,
  iconRightColor,
  iconRightSize,
  onPressIconRight,
  testIDIconRight,
  rightTitle,
  title,
  onPress,
  titleStyle,
  style,
  useDivider,
  useBold,
  rightTitleStyle,
  headerText,
  headerRightText,
  description,
  disabled,
  subtitle,
  rightContent,
  rightTitleSelectable,
}: Props) => {
  const isVerticalStyle = !!(headerText && headerRightText && description);
  const Title = useBold ? Text.Bold : Text;

  return (
    <TouchableOpacity
      disabled={disabled}
      activeOpacity={0.7}
      onPress={onPress}
      testID={testID}
      style={[styles.container, style, isVerticalStyle && styles.containerRow]}
      accessibilityLabel={testID || accessibilityLabel}
    >
      {isVerticalStyle &&
        (imageLeft ? (
          <View style={styles.contentImage}>
            <FastImage
              resizeMode={FastImage.resizeMode.contain}
              style={styles.imageItem}
              source={imageLeft as number}
            />
          </View>
        ) : (
          <View style={styles.contentImage}>
            <Icon
              name={iconLeft}
              style={styles.iconLeft}
              size={22}
              color={iconLeftColor as string}
            />
          </View>
        ))}
      <View style={[styles.content, isVerticalStyle && styles.containerColumn]}>
        {isVerticalStyle && (
          <View style={styles.rightTextContainer}>
            <Text style={titleStyle} testID={`title-${testID}`}>
              {headerText}
            </Text>
            <Text style={styles.headerRightText} testID={`subtitle-${testID}`}>
              {headerRightText}
            </Text>
          </View>
        )}
        {!isVerticalStyle && (
          <Icon
            size={18}
            style={styles.iconContainer}
            color={iconLeftColor as string}
            name={iconLeft}
          />
        )}
        {!isVerticalStyle && imageLeft !== 1 && (
          <View style={styles.iconContainer}>
            <FastImage
              resizeMode={FastImage.resizeMode.contain}
              style={styles.imageItem}
              source={imageLeft as number}
            />
          </View>
        )}
        <View style={styles.listContent}>
          <View style={rightContent ? styles.innerListContentFreeStyle : styles.innerListContent}>
            <View style={styles.leftTextContainer}>
              <View>
                {typeof title === 'string' ? (
                  <Title testID={testIDTitle || `title-${testID}`} style={titleStyle}>
                    {title}
                  </Title>
                ) : (
                  title
                )}
                {subtitle}
              </View>
            </View>
            {rightContent || (
              <View style={styles.rightTextContainer}>
                {(!!description || !!rightTitle) && (
                  <Text
                    testID={testIDRightTitle}
                    style={rightTitleStyle}
                    selectable={rightTitleSelectable}
                  >
                    {isVerticalStyle ? description : rightTitle}
                  </Text>
                )}
                {iconRight ? (
                  <Icon
                    style={styles.iconRight}
                    size={iconRightSize as number}
                    name={iconRight}
                    color={iconRightColor as string}
                    onPress={onPressIconRight}
                    testID={testIDIconRight}
                  />
                ) : undefined}
              </View>
            )}
          </View>
        </View>
      </View>
      {useDivider && <Divider color={Colors.GRAYE0} />}
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    justifyContent: 'center',
    paddingTop: 20,
  },
  containerRow: {
    ...AppStyles.rowItems,
    justifyContent: 'flex-start',
  },
  containerColumn: {
    flex: 1,
    flexDirection: 'column',
  },
  content: {
    flexDirection: 'row',
  },
  contentImage: {
    justifyContent: 'center',
    paddingRight: 15,
  },
  leftTextContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    flex: 1,
  },
  rightTextContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  iconContainer: {
    marginRight: 10,
  },
  imageItem: {
    width: 36,
    height: 36,
  },
  iconLeft: {
    marginRight: 15,
  },
  iconRight: {
    marginLeft: 15,
  },
  headerRightText: {
    color: Colors.GRAY,
    fontSize: 12,
  },
  listContent: {
    flex: 1,
  },
  innerListContent: {
    ...AppStyles.rowItemsCenterSpace,
    marginBottom: 15,
  },
  innerListContentFreeStyle: {
    ...AppStyles.rowItemsSpace,
    marginBottom: 15,
  },
});

ButtonMenuList.defaultProps = {
  accessibilityLabel: undefined,
  testIDTitle: undefined,
  testIDRightTitle: undefined,
  iconLeftColor: Colors.GREEN_47,
  iconRightColor: Colors.GREEN_47,
  iconRightSize: 12,
  iconLeft: undefined,
  imageLeft: 1,
  iconRight: ICON_ARROW_RIGHT,
  onPressIconRight: undefined,
  testIDIconRight: undefined,
  useDivider: true,
  useBold: true,
  headerText: undefined,
  headerRightText: undefined,
  description: undefined,
  rightContent: undefined,
  disabled: false,
  subtitle: undefined,
  rightTitleStyle: {},
  style: {},
  titleStyle: {},
  rightTitle: undefined,
  title: undefined,
  onPress: () => null,
  rightTitleSelectable: false,
};

export default ButtonMenuList;
