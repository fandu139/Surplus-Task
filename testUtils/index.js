import AuthContext from '../App/context/Auth';
import React from 'react';
import { act, fireEvent, render } from '@testing-library/react-native';

const customRender = (ui, options) => {
  const isAuthenticated = options?.isAuthenticated || false;
  const renderOptions = options?.rendererOptions;
  const authContext = options?.authContext;
  return render(
    <AuthContext.Provider value={{ isAuthenticated, ...authContext }}>{ui}</AuthContext.Provider>,
    renderOptions,
  );
};

// override render method
export { customRender as render };

export function wait(ms) {
  return act(() => new Promise((r) => setTimeout(r, ms)));
}

// Due to debounce on Button UI Kit, we need to add milisecond waiting timer
export function fireAsyncPress(element, ms = 100) {
  fireEvent.press(element);
  return wait(ms);
}
