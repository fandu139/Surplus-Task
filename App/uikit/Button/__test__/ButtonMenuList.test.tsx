import ButtonMenuList from '..//ButtonMenuList';
import React from 'react';
import { render } from '../../../../testUtils';

describe('Button Menu List component test', () => {
  describe('snapshot test', () => {
    it('should render correctly', () => {
      const tree = render(
        <ButtonMenuList
          testID="button-menu-list-default"
          title="Test Button Menu List"
          onPress={() => null}
        />,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
