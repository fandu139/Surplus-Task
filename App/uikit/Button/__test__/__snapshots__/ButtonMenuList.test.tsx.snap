// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Menu List component test snapshot test should render correctly 1`] = `
<View
  accessibilityLabel="button-menu-list-default"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "justifyContent": "center",
      "opacity": 1,
      "paddingTop": 20,
    }
  }
  testID="button-menu-list-default"
>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        false,
      ]
    }
  >
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 15,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
            }
          }
        >
          <View>
            <Text
              accessibilityLabel="title-button-menu-list-default"
              allowFontScaling={false}
              style={
                [
                  {
                    "color": "#202020",
                    "fontSize": 14,
                    "fontWeight": "normal",
                  },
                  {},
                  undefined,
                  undefined,
                ]
              }
              testID="title-button-menu-list-default"
            >
              Test Button Menu List
            </Text>
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <View
            accessibilityLabel="button-icon"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": true,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={false}
            hitSlop={
              {
                "bottom": 15,
                "left": 15,
                "right": 15,
                "top": 15,
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="button-icon"
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": undefined,
                    "fontSize": 12,
                  },
                  {
                    "color": "#47bba5",
                    "marginLeft": 15,
                  },
                  {
                    "fontFamily": "Icon",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              Ó§è
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "backgroundColor": "#E0E0E0",
        "height": 1,
      }
    }
  />
</View>
`;
